contrib/pam_zfs_key/pam_zfs_key.c

```diff
@@ -0,0 +1,737 @@
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the <organization> nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Copyright (c) 2020, Felix Dörre
+ * All rights reserved.
+ */
+
+#include <sys/dsl_crypt.h>
+#include <sys/byteorder.h>
+#include <libzfs.h>
+
+#include <syslog.h>
+
+#include <sys/zio_crypt.h>
+#include <openssl/evp.h>
+
+#define	PAM_SM_AUTH
+#define	PAM_SM_PASSWORD
+#define	PAM_SM_SESSION
+#include <security/pam_modules.h>
+
+#if	defined(__linux__)
+#include <security/pam_ext.h>
+#elif	defined(__FreeBSD__)
+#include <security/pam_appl.h>
+void
```

Comment:
Looks like this needs to be declared static.
And do you want to make this function do something?

Reply:
Yes, I wanted to wait for the other CI jobs to complete to resolve this all in one go. Currently I think it is ok to disable the log messages on freebsd, but if you know of a good location where they should go, I am open to suggestions.

Reply:
They could go to syslog. Here is an implementation of pam_syslog in FreeBSD's login program. If you can come up with something that would work in both places I can probably get a pam_ext.h added and commit your pam_syslog to FreeBSD so this can be simplified in the future.

Reply:
The problem is that they currently have completely different signatures. What about this implementation:
```
static void
pam_syslog(pam_handle_t *pamh, int loglevel, const char *fmt, ...)
{
	va_list args;
	va_start(args, fmt);
	vsyslog(loglevel, fmt, args);
	va_end(args);
}
static void
pam_perror(const char *msg)
{
	pam_syslog(NULL, LOG_ERR, "%s: %s", msg, pam_strerror(pamh, pam_err))
}
```
and changing the `pam_syslog` calls in `login.c` to `pam_perror`? Is that something we could do, to harmonize the interface? I hope that `__BSD_VISIBLE` is declared in the relevant contexts, so [`vsyslog`](https://github.com/freebsd/freebsd/blob/1d6e4247415d264485ee94b59fdbc12e0c566fd0/sys/sys/syslog.h#L199) is available.

Reply:
That looks reasonable to me! I'll see about getting it upstreamed to FreeBSD, meanwhile feel free to incorporate your pam_syslog implementation in this PR. I'd maybe add an include for stdarg.h to be explicit, and keep an eye out for the style bot (`make checkstyle` if you want to check locally), it may want a blank line between the local variables and the rest of the function body.

Reply:
Actually I found `openpam_log(level, fmt, ...)` in `security/openpam.h` already exists for logging to syslog. You may as well use that, even a define to throw away the pam handle would work.

Reply:
`openpam_log` has a small problem: the log levels are not the default system loglevels and it has code to translate it. One would have to use `PAM_LOG_ERROR` instead of `LOG_ERROR`: https://github.com/freebsd/freebsd/blob/1d6e4247415d264485ee94b59fdbc12e0c566fd0/contrib/openpam/lib/libpam/openpam_log.c#L99

Reply:
Ah bummer, good catch. Back to plan A then...

Reply:
I'm a bit curious what that pam handle argument is supposed to be for 🤔

Reply:
On linux, this is used to print out some context information with the log line: https://github.com/linux-pam/linux-pam/blob/0e9b286afe1224b91ff00936058b084ad4b776e4/libpam/pam_syslog.c#L81